DOC: Add `>>> import numpy as np` stubs everywhere

agriyakhetarpal opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->


## Description

This PR is a precursor to gh-26745. It adds the `>>> import numpy as np` statement in the docstrings and documentation-based examples everywhere wherever applicable to enable the use of the NumPy API under the `np` namespace. This aims to make examples self-contained, which will be helpful towards being able to run them in isolated mini-notebooks using [Pyodide](https://github.com/pyodide/pyodide/tree/main/packages/numpy) as created by the [`TryExamples` directive](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/try_examples.html) via the [`jupyterlite-sphinx` extension for Sphinx](https://jupyterlite-sphinx.readthedocs.io/en/stable/).

### Changes made

1. The addition of `>>> import numpy as np` statements in most files (certain files like `_npyio_impl.py` have been left out, please let me know if that file should be modified – it won't run well in Pyodide because of the lack of a filesystem to access, but the examples can be made self-contained in general and outside of Pyodide)
3. General extra whitespace removals, changes to punctuation, and a few updates to grammar here and there

## Additional context

- This change was requested in https://github.com/numpy/numpy/pull/26745#pullrequestreview-2138930610
- This is where such a change was discussed for SciPy: https://github.com/scipy/scipy/issues/13049/
- SciPy's interactive examples PR: https://github.com/scipy/scipy/pull/20019/

agriyakhetarpal committed:
DOC, TST: Add `>>> import numpy as np` stub to docstrings

agriyakhetarpal committed:
DOC: Add `>>> import numpy as np` to API reference examples

[skip azp] [skip cirrus]

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/_core/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/lib/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/ma/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/linalg/_linalg.py`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/fft/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/random/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/polynomial/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/exceptions.py`

rgommers commented:
Thanks @agriyakhetarpal. Note that the doc build CI is failing. Apologies in advance: I suggested splitting this off from your previous PR, but won't have time to actually review in detail over the next days.

Request to potential reviewers: this PR has the potential for a lot of bikeshedding. In case you are feeling an inclination to actually start such a process, I strongly recommend to think twice before doing that, and read the linked threads under "additional context" first.

ngoldbaum commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -409,24 +409,27 @@ alias for "matrix "in NumPy.
 
->>> a = np.asmatrix('1 2 3; 4 5 3')
->>> print((a*a.T).I)
+  >>> import numpy as np
```
You only need to do this once per `rst` file or example section in a docstring.

Also, to ease reviewing, I think it would be best to only add imports and not make language or other changes in this PR.

agriyakhetarpal replied:
> You only need to do this once per rst file or example section in a docstring.

Actually, we'll need to do this per block of code, unless we want all Examples 1–3 here under the same `TryExamples` block. I think it depends on the length of the content present between each block. For example, the imports would be good to have just once here, however, in other places where there is a lot of text between code blocks, it makes the examples a tad difficult to follow (especially because of the limited width available inside the notebook layout, plus considering the primary and secondary sidebars take up a lot of space).

> Also, to ease reviewing, I think it would be best to only add imports and not make language or other changes in this PR.

Sure – I can leave the ones that I already included, but will make note of this and won't add any further changes! Please let me know if I should remove them altogether.

ngoldbaum replied:
Oops, you're right. I thought SciPy stuck to one import per rst file, but that's wrong on a second look.

agriyakhetarpal replied:
Added the SciPy implementation to the PR description as well for other readers

ngoldbaum commented on doc/source/reference/arrays.dtypes.rst:
```diff
@@ -401,19 +425,21 @@ Type strings
 
-       Data type with fields ``r``, ``g``, ``b``, ``a``, each being
-       an 8-bit unsigned integer:
+         >>> import numpy as np
```
IMO these new imports should appear along with the code right before they're needed, not in a new code block preceding any text that's already there.

agriyakhetarpal replied:
Thank you, addressed in 811830b955a617ff5dddc4df5f3fdd85e94aca31 (sorry for the force-push, forgot to skip Azure Pipelines and CirrusCI).

agriyakhetarpal committed:
DOC: Make new imports succeed text in examples

[skip azp] [skip cirrus]

agriyakhetarpal commented:
Not sure if I understand the "Diverging branches can't be fast-forwarded" error in CircleCI with the 128 exit code. My branch and my commits seem fine, so it's most likely a downtime incident with GitHub.

ngoldbaum commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -625,13 +633,14 @@ As mentioned previously, the flat attribute of ndarray objects returns
-10 20
-15 25
-20 30
+  >>> import numpy as np
```
this example doesn't need the import, does it?

ngoldbaum replied:
And please double-check all the imports you're adding are needed.

agriyakhetarpal replied:
Thanks, I did try my best to make sure I didn't add any unnecessary imports – will make note of this. In this particular example: I think we need it because `a` as an array isn't defined (it's defined in the "Default iteration" heading above it, though. I added `a` from to make the example work in dba775253d6b82d8e565e272ee1d83267264f2b2

ngoldbaum replied:
So this is part of what I was alluding to above with the one import per rst thing, the examples can be interspersed with text but still share a namespace and still make sense to the reader IMO. That means an object declared earlier in the same example is available later without needing to be redefined in every example, which IMO makes the docs harder to read. I guess that makes it harder to make the examples runnable with pyodide?

agriyakhetarpal replied:
> I guess that makes it harder to make the examples runnable with pyodide?

Yes, in some way. However, there are other directives offered by `jupyterlite-sphinx` such as the [`NotebookLite`](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/notebooklite.html) directive that can convert an entire page to a runnable example (if it is a MyST-style notebook, for example – and such conversion demands slightly more work). https://github.com/scipy/scipy/pull/20303 and https://github.com/scipy/scipy/pull/21042 are some references for this.

> the examples can be interspersed with text but still share a namespace and still make sense to the reader IMO

For examples that are supposed with the `TryExamples` directive, this could cause problems, because `.. try_examples::` stops the collection of doctest-based code and text once it hits a change in indentation, i.e., after that one needs to add the directive again for another set of code in `>>>`. This is probably okay for the API examples, because the "Examples" rubric is at the end of the docstring with no extra text after it. However, these sections will probably need the imports so if we have to truly make the examples" truly self-contained". I agree that it could probably make following along a little difficult and doing so isn't one of the _primary_ goals of this PR...

agriyakhetarpal committed:
DOC: Define ndarray object `a` for "Flat iteration" example

agriyakhetarpal committed:
Merge branch 'main' into add-import-numpy-as-np-statement

agriyakhetarpal committed:
DOC: Reorder stubs for a few polynomial docstrings

[skip azp] [skip cirrus]

agriyakhetarpal commented:
There was an incident for GitHub repo clones on Friday, which seems to have been resolved today, but there's a different error in CircleCI from `conf.py` with the line: `assert 'numpy._core._add_newdocs_scalars' not in sys.modules`. I can't reproduce this locally with `spin docs --clean` – I'll keep investigating.

agriyakhetarpal committed:
Merge branch 'main' into add-import-numpy-as-np-statement

agriyakhetarpal committed:
DOC: Make "Copies and views" self-contained

[skip azp] [skip cirrus]

agriyakhetarpal committed:
DOC: Make "Array creation" self-contained

[skip azp] [skip cirrus]

agriyakhetarpal committed:
DOC: Make "Broadcasting" self-contained

[skip azp] [skip cirrus]

agriyakhetarpal commented:
The docs build has been fixed in `main`, I updated my branch and I added a few commits for the user guide. Many of the rest of the pages under the user guide are better poised to be converted to Jupytext-based notebooks as discussed in https://github.com/numpy/numpy/pull/26745#issue-2360240687 (under the "Key changes" section).

ngoldbaum commented on doc/source/reference/arrays.scalars.rst:
```diff
@@ -191,29 +191,31 @@ Inexact types
+      >>> f16 == f32 == f64
+      False
+      >>> f16, f32, f64
+      (0.1, 0.1, 0.1)
```
With the indentation change the text now shows up as "raw" text, which I don't think we want. If the only way to avoid the import in the second example is to do this, I'd rather have the extra import.

agriyakhetarpal replied:
Thanks for noticing! I fixed this based on your other comment in 30301ed484be5a22a0b2e756ea09afc32159cf38.

ngoldbaum commented on doc/source/reference/arrays.scalars.rst:
```diff
@@ -191,29 +191,31 @@ Inexact types
    This means that variables with equal binary values but whose datatypes are of
    different precisions may display differently::
```
Although maybe it's just the use of `::` here that's problematic.

agriyakhetarpal replied:
Yes, I noticed this issue while working on #26745 in many places. Would it be better to handle such changes here? It would reduce the diff there, but I think there's no way to test the rendering of the examples out without having integrated `jupyterlite-sphinx`, so we might as well test it there better at the expense of a larger diff.

ngoldbaum replied:
If the plan is to merge this PR and then that one, I'd prefer to not knowingly make some examples not render very nicely. Just in case they get missed or that PR takes a long time to land for whatever reason. I don't mind making the diff a bit longer to sort out that, I think for this specific example all that's needed is to delete the `::` and it will automatically get the nice example markup you changed the docs to use elsewhere in this PR.

ngoldbaum reviewed (commented):
I really like the new example blocks that show up if you adjust the indentation for long examples! I think there's just one where it's not rendering correctly, see inline comment.

Overall I really like this change and distributed over the entire community I think this will save a lot of human time where currently people need to add the import after copy/pasting code from the docs and seeing an error.

ngoldbaum commented:
<img width="541" alt="image" src="https://github.com/numpy/numpy/assets/3126246/cca28fd4-4925-48c9-8187-8e829fb39ace">

Compare with: https://numpy.org/doc/stable/reference/random/index.html

agriyakhetarpal committed:
DOC: Fix indentation for "Inexact types" example

agriyakhetarpal committed:
DOC: Revert `default_rng` example text indentation

[skip azp] [skip cirrus]

agriyakhetarpal commented:
> Compare with: [numpy.org/doc/stable/reference/random/index.html](https://numpy.org/doc/stable/reference/random/index.html)

I restored this change to the original in 25f9a99e2db092356b8392d48c92004949efdda6. It's not neat, but adding the `TryExamples` directive will fix this indentation problem.

mattip commented:
There is a merge conflict, I guess that is to be expected in a PR this big. Perhaps we could push this over the finish line, and fix any lingering indentation problems in follow-on PRs?

agriyakhetarpal committed:
Merge branch 'main' into add-import-numpy-as-np-statement

agriyakhetarpal commented:
Thanks, @mattip – I resolved the merge conflicts, and this is ready for review again. I, too, suggest fixing indentation problems in follow-on PRs.

mattip commented:
Thanks @agriyakhetarpal

mattip commented:
I wonder if there is a regex we could use on the rendered documentation to find indentation problems? Looking at the [random example with extra indentation](https://output.circle-artifacts.com/output/job/9c92a67a-ce63-49ec-9525-bd9db8642f41/artifacts/0/doc/build/html/reference/random/index.html) it seems the problem is that a `<div class="doctest highlight-default notranslate"> ... </div>` code block is contained inside a `<blockquote>...</blockquote>` block. Can we search for excessive use of `blockquote`?

agriyakhetarpal commented:
This is a good idea, I can look into doing this. With this PR merged, I'll resume my work on gh-26745 and fix any cases of improper indentation.
